### DISCUSSION TITLE
Ignoring option install-extension: not supported for code-insiders

### DISCUSSION THREAD

**Discussion by sobolevn** [community/community#27138] [Codespaces]
<p>I am trying to run this file: <a href="https://github.com/sobolevn/dotfiles/blob/ae5e8d992e9479793d86c232ea82deb66b1e3f2a/vscode/plugins.sh" class="inline-onebox" rel="noopener nofollow ugc">dotfiles/plugins.sh at ae5e8d992e9479793d86c232ea82deb66b1e3f2a · sobolevn/dotfiles · GitHub</a></p>
<p>It contains several <code>sh</code> lines like <code>code --install-extension teabyii.ayu</code>,<br>
but it fails to be executed in Codespaces.</p>
<p>Error:</p>
<pre><code class="lang-auto">root@codespaces_76fb96:~/dotfiles# code --install-extension teabyii.ayu
Ignoring option install-extension: not supported for code-insiders.
At least one file or folder must be provided.
</code></pre>
<p>What are the possible solutions here?</p>

**Comment by jkeech** [comment, marked as answer]
<p>Hi <a class="mention" href="/sobolevn">@sobolevn</a>, installing extensions as part of dotfiles poses some challenges, since dotfiles install runs in parallel with user connection, and there’s no guarantee you are connected by the time it runs, or even which VS Code version you are connecting from (insiders, stable, etc). Without knowing which VS Code version you are connecting from, you cannot know which version to install the extension into.</p>
<p>There are two alternative recommended methods for installing extensions automatically:</p>
<ol>
<li>Per-user extensions can be installed via settings sync in VS Code, which is enabled automatically in Codespaces. This will have a similar effect to dotfiles: they will be installed in all of your codespaces.</li>
<li>Per-repo extensions can be specified in the devcontainer.json for a given repo. Then every user who creates a codespace in that repo will get that extension.</li>
</ol>
<p>I would use (1) for extensions that are mostly personal preference like themes and other UI-related extensions, and (2) for things like language extensions, SDKs, debuggers, etc which are needed to build and run the project, and which everyone in that repo is going to need.</p>
<p>Hope that helps!</p>

**Comment by sobolevn** [comment]
<p>Thanks for your help!</p>
<p>Settings sync kinda works, but some minor things are not clear:</p>
<ol>
<li>If I have to enable it everytime I spawn a new Codespace instance (cannot find any docs references on this one)</li>
<li>I cannot modify settings in the “build time”. I had a plan to drop and modify some keys during the <code>dotfiles</code> installation process. For example, I am going to use light theme in Codespaces and a dark theme locally. Does settings sync support settings substitution?</li>
</ol>

**Comment by jkeech** [comment]
<p>For (1), you shouldn’t need to do anything to enable settings sync. It’s on automatically in Codespaces. Let us know if you are seeing something different.</p>
<p>For (2), I don’t think that’s supported currently. There’s not a great way to have Codespaces-specific config that is applied to all codespaces and not roamed back to desktop. cc <a class="mention" href="/sandy081">@sandy081</a> who works on settings sync for VS Code.</p>

**Comment by sobolevn** [comment]
<blockquote>
<p>installing extensions as part of dotfiles poses some challenges, since dotfiles install runs in parallel with user connection, and there’s no guarantee you are connected by the time it runs</p>
</blockquote>
<p>Might be a good idea to add some hooks like <code>post_install.sh</code>, where we would have access to the existing <code>vscode</code> install.</p>
